modules/inject/src/main/scala/io/janstenpickle/trace4cats/inject/ContextConstructor.scala

```diff
@@ -0,0 +1,51 @@
+package io.janstenpickle.trace4cats.inject
+
+import cats.Applicative
+import cats.effect.Resource
+
+trait ContextConstructor[F[_], R, R0] { self =>
+  def F: Applicative[F]
+
+  def apply(r: R): Resource[F, R0] = resource(r)
+
+  def resource(r: R): Resource[F, R0]
+
+  def mapBoth[R1](f: (R, R0) => R1): ContextConstructor[F, R, R1] = new ContextConstructor[F, R, R1] {
+    override def F: Applicative[F] = self.F
+
+    override def resource(r: R): Resource[F, R1] = self.resource(r).map(f(r, _))(F)
+  }
+
+  def mapResult[R1](f: R0 => R1): ContextConstructor[F, R, R1] = mapBoth((_, r0) => f(r0))
+
+  def map[R1](f: R => R1): ContextConstructor[F, R, R1] = mapBoth((r, _) => f(r))
+
+  def contramap[R1](f: R1 => R): ContextConstructor[F, R1, R0] = new ContextConstructor[F, R1, R0] {
+    override def F: Applicative[F] = self.F
+
+    override def resource(r: R1): Resource[F, R0] = self.resource(f(r))
+  }
+
+  def mapBothF[R1](f: (R, R0) => F[R1]): ContextConstructor[F, R, R1] = new ContextConstructor[F, R, R1] {
```

Comment:
I would replace `mapBoth`, `mapBothF`, `mapBothResource` with just one method:
```scala
def withInput[In1 <: In]: ContextConstructor[F, In1, (In1, Out)]
```

This allows to chain the result with the previously defined methods.